packages/mermaid/src/diagrams/xychart/parser/xychart.jison

```diff
@@ -0,0 +1,201 @@
+["]                                       this.begin("string");
+<string>["]                               this.popState();
+<string>[^"]*                             return "STR";
+
+
+[\[]                                      return 'SQUARE_BRACES_START'
+[\]]                                      return 'SQUARE_BRACES_END'
+[A-Za-z]+                                 return 'ALPHA';
+":"                                       return 'COLON';
+\+                                        return 'PLUS';
+","                                       return 'COMMA';
+"="                                       return 'EQUALS';
+\=                                        return 'EQUALS';
+"*"                                       return 'MULT';
+\#                                        return 'BRKT';
+[\_]                                      return 'UNDERSCORE';
+"."                                       return 'DOT';
+"&"                                       return 'AMP';
+\-                                        return 'MINUS';
+[0-9]+                                    return 'NUM';
+\s                                        /* skip */
+";"                                       return 'SEMI';
+[!"#$%&'*+,-.`?\\_/]                      return 'PUNCTUATION';
+<<EOF>>                                   return 'EOF';
+
+/lex
+
+%start start
+
+%% /* language grammar */
+
+start
+  : eol start
+  | directive start
+  | XYCHART chartConfig start
+	| XYCHART start
+  | document
+	;
+
+chartConfig
+  : CHART_ORIENTATION                     { yy.setOrientation($1); }
+  ;
+
+document
+	: /* empty */
+	| document statement
+	;
+
+line
+	: statement
+	;
+
+statement
+  : statement eol
+  | title title_value                     { yy.setDiagramTitle($title_value.trim()); }
+  | X_AXIS parseXAxis
+  | Y_AXIS parseYAxis
+  | LINE parseLineData                    { yy.setLineData({text: '', type: 'text'}, $parseLineData); }
+  | LINE text parseLineData               { yy.setLineData($text, $parseLineData); }
+  | BAR parseBarData                      { yy.setBarData({text: '', type: 'text'}, $parseBarData); }
+  | BAR text parseBarData                 { yy.setBarData($text, $parseBarData); }
+	;
+
+parseLineData
+  : SQUARE_BRACES_START NUMBER_WITH_DECIMAL parseLineData   {$parseLineData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseLineData}
+  | COMMA NUMBER_WITH_DECIMAL parseLineData                 {$parseLineData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseLineData;}
+  | SQUARE_BRACES_END                                       {$$ = []}
+  ;
+
+parseBarData
+  : SQUARE_BRACES_START NUMBER_WITH_DECIMAL parseBarData    {$parseBarData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseBarData}
+  | COMMA NUMBER_WITH_DECIMAL parseBarData                  {$parseBarData.unshift(Number($NUMBER_WITH_DECIMAL)); $$ = $parseBarData;}
+  | SQUARE_BRACES_END                                       {$$ = []}
+  ;
```

Comment:
1. The same problem of unbalanced brackets.
2. This construction repeats `parseLineData`, so both of them `parseLineData` and  `parseBarData` can be combined into one non-term called `data` (or any other, name choices is up to you)
    e.g
    ```
    statement
      : LINE data
      : BAR data
      ;
    data: "[" values "]";
    values: value | "," values | ;
    ```
    If you want values not to be empty:
    ```
    values: value values_list;
    values_list: "," value | values_list | ;
    ```
    Furthermore, the fact that you are [starting the same directive](https://github.com/mermaid-js/mermaid/pull/4413/files#diff-532cbb4afcc95e6caafa23852a6a0edbf55849150d219ec1bb9f11fe763afef6R52-R53) called `data` for `LINE` and `BAR` only emphasizes the assumption that the latter tho must be combined.

Reply:
Done